linkerd/app/inbound/src/detect.rs

```diff
@@ -98,61 +96,74 @@ impl<N> Inbound<N> {
-                            let permit = policy.check_authorized(t.param(), &tls)?;
-                            return Ok(svc::Either::B(Forward::mk(&t, tls, permit)));
+                        if protocol == Protocol::Tls {
+                            return Ok(svc::Either::B(tls));
                         }
 
-                        Ok(svc::Either::A(Tls {
-                            client_addr: t.param(),
-                            orig_dst_addr: t.param(),
-                            status: tls,
-                            policy,
-                        }))
+                        Ok(svc::Either::A(tls))
                     },
                     svc::stack(forward.clone())
                         .push_on_service(svc::MapTargetLayer::new(io::BoxedIo::new))
+                        .push_map_target(Forward::from)
+                        .push(policy::NewAuthorizeTcp::layer(rt.metrics.tcp_authz.clone()))
+                        .check_new_service::<Tls, _>()
                         .into_inner(),
                 )
+                .check_new_service::<(tls::ConditionalServerTls, T), _>()
                 .push(tls::NewDetectTls::layer(TlsParams {
                     timeout: tls::server::Timeout(detect_timeout),
                     identity: rt.identity.clone(),
                 }))
+                .check_new_service::<T, I>()
                 .push_switch(
                     // If this port's policy indicates that authentication is not required and
                     // detection should be skipped, use the TCP stack directly.
-                    |t: T| -> Result<_, DeniedUnauthorized> {
+                    |t: T| -> Result<_, Infallible> {
                         let policy: AllowPolicy = t.param();
                         if policy.protocol() == Protocol::Opaque {
-                            let permit = policy.check_authorized(t.param(), &TLS_PORT_SKIPPED)?;
-                            return Ok(svc::Either::B(Forward::mk(&t, TLS_PORT_SKIPPED, permit)));
+                            const TLS_PORT_SKIPPED: tls::ConditionalServerTls =
+                                tls::ConditionalServerTls::None(tls::NoServerTls::PortSkipped);
+                            return Ok(svc::Either::B(Tls {
+                                client_addr: t.param(),
+                                orig_dst_addr: t.param(),
+                                status: TLS_PORT_SKIPPED,
+                                policy,
+                            }));
                         }
                         Ok(svc::Either::A(t))
                     },
                     svc::stack(forward)
                         .push_on_service(svc::MapTargetLayer::new(io::BoxedIo::new))
+                        .push_map_target(Forward::from)
+                        .push(policy::NewAuthorizeTcp::layer(rt.metrics.tcp_authz.clone()))
+                        .check_new_service::<Tls, I>()
                         .into_inner(),
```

Comment:
very minor nit, not a blocker: seems like there could be a `let` binding for this and we could clone it into each `push_switch`...or is there some type difference i'm missing?